Kernel/Graphics/VirtIOGPU/VirtIOGPU.h

```diff
@@ -0,0 +1,196 @@
+
+#define VIRTIO_GPU_MAX_SCANOUTS 16
+
+#define CONTROLQ 0
+#define CURSORQ 1
+
+namespace Kernel::Graphics {
+
+enum class VirtIOGPUCtrlType : u32 {
+    /* 2d commands */
+    VIRTIO_GPU_CMD_GET_DISPLAY_INFO = 0x0100,
+    VIRTIO_GPU_CMD_RESOURCE_CREATE_2D,
+    VIRTIO_GPU_CMD_RESOURCE_UNREF,
+    VIRTIO_GPU_CMD_SET_SCANOUT,
+    VIRTIO_GPU_CMD_RESOURCE_FLUSH,
+    VIRTIO_GPU_CMD_TRANSFER_TO_HOST_2D,
+    VIRTIO_GPU_CMD_RESOURCE_ATTACH_BACKING,
+    VIRTIO_GPU_CMD_RESOURCE_DETACH_BACKING,
+    VIRTIO_GPU_CMD_GET_CAPSET_INFO,
+    VIRTIO_GPU_CMD_GET_CAPSET,
+    VIRTIO_GPU_CMD_GET_EDID,
+
+    /* cursor commands */
+    VIRTIO_GPU_CMD_UPDATE_CURSOR = 0x0300,
+    VIRTIO_GPU_CMD_MOVE_CURSOR,
+
+    /* success responses */
+    VIRTIO_GPU_RESP_OK_NODATA = 0x1100,
+    VIRTIO_GPU_RESP_OK_DISPLAY_INFO,
+    VIRTIO_GPU_RESP_OK_CAPSET_INFO,
+    VIRTIO_GPU_RESP_OK_CAPSET,
+    VIRTIO_GPU_RESP_OK_EDID,
+
+    /* error responses */
+    VIRTIO_GPU_RESP_ERR_UNSPEC = 0x1200,
+    VIRTIO_GPU_RESP_ERR_OUT_OF_MEMORY,
+    VIRTIO_GPU_RESP_ERR_INVALID_SCANOUT_ID,
+    VIRTIO_GPU_RESP_ERR_INVALID_RESOURCE_ID,
+    VIRTIO_GPU_RESP_ERR_INVALID_CONTEXT_ID,
+    VIRTIO_GPU_RESP_ERR_INVALID_PARAMETER,
+};
+
+struct VirtIOGPUCtrlHeader {
+    u32 type;
+    u32 flags;
+    u64 fence_id;
+    u32 context_id;
+    u32 padding;
+};
+
+struct VirtIOGPURect {
+    u32 x;
+    u32 y;
+    u32 width;
+    u32 height;
+};
+
+struct VirtIOGPURespDisplayInfo {
+    VirtIOGPUCtrlHeader header;
+    struct VirtIOGPUDisplayOne {
+        VirtIOGPURect rect;
+        u32 enabled;
+        u32 flags;
+    } scanout_modes[VIRTIO_GPU_MAX_SCANOUTS];
+};
+
+enum class VirtIOGPUFormats : u32 {
+    VIRTIO_GPU_FORMAT_B8G8R8A8_UNORM = 1,
+    VIRTIO_GPU_FORMAT_B8G8R8X8_UNORM = 2,
+    VIRTIO_GPU_FORMAT_A8R8G8B8_UNORM = 3,
+    VIRTIO_GPU_FORMAT_X8R8G8B8_UNORM = 4,
+
+    VIRTIO_GPU_FORMAT_R8G8B8A8_UNORM = 67,
+    VIRTIO_GPU_FORMAT_X8B8G8R8_UNORM = 68,
+
+    VIRTIO_GPU_FORMAT_A8B8G8R8_UNORM = 121,
+    VIRTIO_GPU_FORMAT_R8G8B8X8_UNORM = 134,
+};
+
+struct VirtIOGPUResourceCreate2D {
+    VirtIOGPUCtrlHeader header;
+    u32 resource_id;
+    u32 format;
+    u32 width;
+    u32 height;
+};
+
+struct VirtIOGPUSetScanOut {
+    VirtIOGPUCtrlHeader header;
+    VirtIOGPURect rect;
+    u32 scanout_id;
+    u32 resource_id;
+};
+
+struct VirtIOGPUMemEntry {
+    u64 address;
+    u32 length;
+    u32 padding;
+};
+
+struct VirtIOGPUResourceAttachBacking {
+    VirtIOGPUCtrlHeader header;
+    u32 resource_id;
+    u32 num_entries;
+    // TODO: Better way of doing this
+    VirtIOGPUMemEntry entries[0];
+};
+
+struct VirtIOGPUTransferToHost2D {
+    VirtIOGPUCtrlHeader header;
+    VirtIOGPURect rect;
+    u64 offset;
+    u32 resource_id;
+    u32 padding;
+};
+
+struct VirtIOGPUResourceFlush {
+    VirtIOGPUCtrlHeader header;
+    VirtIOGPURect rect;
+    u32 resource_id;
+    u32 padding;
+};
+
+class VirtIOGPU final : public VirtIODevice
+    , public RefCounted<VirtIOGPU> {
```

Comment:
We usually format that like this:
```suggestion
class VirtIOGPU final
    : public VirtIODevice
    , public RefCounted<VirtIOGPU> {
```

Reply:
Will fix in next revision.